File: DSA/Hima-Swarupa-Padala/day13q1.py
def seq_dict_words(string, dictionary):

    n = len(string)
 
    dp = [False] * (n + 1)
    dp[0] = True 

    for i in range(1, n + 1):
        for j in range(i):
            
            if dp[j] and string[j:i] in dictionary:
                dp[i] = True
                break

    if dp[n]:
        return "YES"
    
    else:
        return "NO"

string = "applepie"
dictionary = {"apple", "pie"}
print(seq_dict_words(string, dictionary))  

File: DSA/Ishita-Nigam/Day13/Day13_Q1.java
import java.util.HashSet;
import java.util.Set;

public class Day13_Q1 {
    public static void main(String[] args) {
        String string = "applepie";
        String[] dictionaryArray = { "apple", "pie" };

        Set<String> dictionary = new HashSet<>();
        for (String word : dictionaryArray) {
            dictionary.add(word);
        }

        if (canSegmentString(string, dictionary)) {
            System.out.println("YES");
        } else {
            System.out.println("NO");
        }
    }

    public static boolean canSegmentString(String s, Set<String> dictionary) {
        boolean[] dp = new boolean[s.length() + 1];
        dp[0] = true;

        for (int i = 1; i <= s.length(); i++) {
            for (int j = 0; j < i; j++) {
                if (dp[j] && dictionary.contains(s.substring(j, i))) {
                    dp[i] = true;
                    break;
                }
            }
        }

        return dp[s.length()];
    }
}
